Homework_11_Java_Circus(Exceptions_Generics_Annotations)/Homework_11_Java_Circus/src/Circus.java

```diff
@@ -0,0 +1,84 @@
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Represents a circus with a list of performers.
+ */
+
+public class Circus<T extends CircusPerformer> {
+	private List<T> performers;
+
+	public Circus() {
+		performers = new ArrayList<>();
+	}
+
+	public void addPerformer(T performer) {
+		performers.add(performer);
+	}
+
+	/**
+	 * Removes a performer from the circus.
+	 */
+	public T removePerformer(T performer) {
+		if (performers != null && performers.remove(performer)) {
+			return performer;
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * Gets a performer by name.
+	 */
+	public T getPerformerByName(String name) {
+		for (T performer : performers) {
+			if (performer.getName().equals(name)) {
+				return performer;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Lists all performers in the circus.
+	 */
+	public List<T> listPerformers() {
+		return new ArrayList<>(performers);
+	}
+
+	/**
+	 * Saves the list of performers to a file.
+	 */
+	public void saveToFile(String filename) {
+		try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
+			for (T performer : performers) {
+				writer.write(performer.getName() + ", " + performer.getAct() + ", " + performer.getExperience());
+				writer.newLine();
+			}
+		} catch (IOException e) {
+			System.out.println("Error saving to file: " + e.getMessage());
+		}
+	}
+
+	/**
+	 * Loads the list of performers from a file.
+	 */
+	public void loadFromFile(String filename) {
+		try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
+			String line;
+			performers.clear();
+			while ((line = reader.readLine()) != null) {
+				String[] parts = line.split(", ");
+				if (parts.length == 3) {
+					String name = parts[0];
+					String act = parts[1];
+					int experience = Integer.parseInt(parts[2]);
+					addPerformer((T) new CircusPerformer(name, act, experience));
```

Comment:
Quite interesting...it forces you to cast to T (which is still safe enough as `T extends CircusPerformer`), looks like the only reason for that is despite `T extends CircusPerformer` , it doesn't explicitly mean T **is always** a CircusPerfomer, that's why compiler have a suspicion here

Reply:
**Make sure you read this after the lesson about functional interfaces and lambdas**

It's a curious situation, I've looked for the avoidance but there's no 100% way -> the only 'workaround' is modifying the method to expect `Function<String[], T> creator`  (which is a functional interface type from java.lang) which allows to pass a lambda expression as a parameter later:

Here the casting will still be executed within a lambda, so we're just "hiding" it although not avoiding + this would make the code much more complex to understand, so your way is the best choice

```java
public void loadFromFile(String filename, Function<String[], T> creator) {
    try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
        String line;
        performers.clear();
        while ((line = reader.readLine()) != null) {
            String[] parts = line.split(", ");
            if (parts.length == 3) {
                addPerformer(creator.apply(parts));
            }
        }
    } catch (IOException e) {
        System.out.println("Error loading from file: " + e.getMessage());
    }
}
```

```java
circus.loadFromFile(filename, parts -> new CircusPerformer(parts[0], parts[1], Integer.parseInt(parts[2])));
```